protocol/bc/txinput.go

```diff
@@ -18,279 +18,233 @@ type (
 
 	TypedInput interface {
 		IsIssuance() bool
+		readWitness(r io.Reader, assetVersion uint64) error
 	}
 )
 
 var errBadAssetID = errors.New("asset ID does not match other issuance parameters")
 
-func (t *TxInput) AssetAmount() AssetAmount {
-	if ii, ok := t.TypedInput.(*IssuanceInput); ok {
-		return AssetAmount{
-			AssetID: ii.AssetID(),
-			Amount:  ii.Amount,
-		}
+func (t *TxInput) writeTo(w io.Writer, serflags uint8) error {
+	_, err := blockchain.WriteVarint63(w, t.AssetVersion)
+	if err != nil {
+		return err
 	}
-	si := t.TypedInput.(*SpendInput)
-	return si.AssetAmount
-}
-
-func (t *TxInput) AssetID() AssetID {
-	if ii, ok := t.TypedInput.(*IssuanceInput); ok {
-		return ii.AssetID()
+	_, err = blockchain.WriteExtensibleString(w, func(w io.Writer) error {
+		return t.WriteInputCommitment(w)
+	})
+	if err != nil {
+		return err
 	}
-	si := t.TypedInput.(*SpendInput)
-	return si.AssetID
-}
-
-func (t *TxInput) Amount() uint64 {
-	if ii, ok := t.TypedInput.(*IssuanceInput); ok {
-		return ii.Amount
+	_, err = blockchain.WriteVarstr31(w, t.ReferenceData)
+	if err != nil {
+		return err
 	}
-	si := t.TypedInput.(*SpendInput)
-	return si.Amount
-}
-
-func (t *TxInput) ControlProgram() []byte {
-	if si, ok := t.TypedInput.(*SpendInput); ok {
-		return si.ControlProgram
+	if serflags&SerWitness != 0 {
+		_, err = blockchain.WriteExtensibleString(w, func(w io.Writer) error {
+			return t.writeInputWitness(w)
+		})
+		if err != nil {
+			return err
+		}
 	}
 	return nil
 }
 
-func (t *TxInput) IssuanceProgram() []byte {
-	if ii, ok := t.TypedInput.(*IssuanceInput); ok {
-		return ii.IssuanceProgram
-	}
-	return nil
-}
+// WriteInputCommitment writes the bare input commitment to w. It's up
+// to the caller to wrap it in an "extensible string" for use in
+// transaction serialization.
+func (t *TxInput) WriteInputCommitment(w io.Writer) (err error) {
+	if t.AssetVersion == 1 {
+		switch inp := t.TypedInput.(type) {
+		case *IssuanceInput:
+			_, err = w.Write([]byte{0}) // issuance type
+			if err != nil {
+				return err
+			}
+			_, err = blockchain.WriteVarstr31(w, inp.Nonce)
+			if err != nil {
+				return err
+			}
+			assetID := inp.AssetID()
+			_, err = w.Write(assetID[:])
+			if err != nil {
+				return err
+			}
+			_, err = blockchain.WriteVarint63(w, inp.Amount)
+			return err
 
-func (t *TxInput) Arguments() [][]byte {
-	switch inp := t.TypedInput.(type) {
-	case *IssuanceInput:
-		return inp.Arguments
-	case *SpendInput:
-		return inp.Arguments
+		case *SpendInput:
+			_, err = w.Write([]byte{1}) // spend type
+			if err != nil {
+				return err
+			}
+			_, err = inp.Outpoint.WriteTo(w)
+			if err != nil {
+				return err
+			}
+			// Nested extensible string
+			_, err = blockchain.WriteExtensibleString(w, func(w io.Writer) error {
+				return inp.writePrevoutCommitment(w, t.AssetVersion)
+			})
+			return err
+		}
+		return fmt.Errorf("unknown input type %T", t.TypedInput)
```

Comment:
Not a conscious one, but your question made me think about the implicit one I've been using, which I think is: if the function doesn't already return an `error` value, don't add one for this purpose (which is the case for e.g. accessors); but if it does, then use it.

Not commenting on whether or not that's the right guideline to be using...